Examples/puzzle.lisp

```diff
@@ -41,7 +41,7 @@
 	  command))))
 
 (define-presentation-type puzzle-cell ()
-  :inherit-from '(integer 1 15))
+  :inherit-from '(integer 0 15))
```

Comment:
`(integer 1 15)` is right. Only cells from 1 to 15 are interactive. You will never see the 0th cell.

Reply:
that's a good example of issues which may be revealed by calling presentation-typep when creating the presentation. 

`puzzle-cell` represents is a cell-id. The puzzle "0" may be presented on a cell with the id 4. Cell with id 0 may be presenting a puzzle with value 8.

Reply:
> puzzle-cell represents is a cell id.

No, it doesn't.

Reply:
``
(updating-output (stream :unique-id cell-id 
				       :cache-value value)
		(formatting-cell (stream :align-x :right)
		  (unless (zerop value)
		    (with-output-as-presentation
			(stream cell-id 'puzzle-cell)
		      (format stream "~2D" value)))))
```

presented object here is` cell-id`, not `value`.

also, see this modified demo screencap which allows inspecting cells: http://turtleware.eu/static/paste/0f8f1506-cells.webm

Reply:
All right, I see now. You have a very fundamental gap in understanding how presentations work.

Any CLIM based program can be divided into the application-logic part and the user-interface part.

`cell-id` here is part of the application-logic. it is the index into the `puzzle-array`.It has type `integer` and its value ranges from 0 to 15. `value` would be the the integer value at any index `cell-id` of `puzzle-array`. Value of `value` too ranges from 0 to 15.

Now cosider the user interface part.

A new presentation-type is defined as `puzzle-cell`. its type is defined as `integer`. ignore its range for now.

`with-output-as-presentation` takes an `stream`, an `object` and a `presentation-type` and creates a presentation of type `presentation-type`. Note that presentation itself is an object (This is the fundamental detail you miss). It stores the underlying object and its presentation type. In this example, `with-output-as-presentation (stream cell-id 'puzzle-cell)` would create a new presentation object of type `puzzle-cell` (of type integer, ignore range for now) with its underlying object `cell-id` (of type integer, ranges from 0 to 15).

Now consider the display function.
```lisp
(defmethod draw-puzzle ((puzzle puzzle) stream &key max-width max-height)
  (declare (ignore max-width max-height))
  (let ((puzzle-array (puzzle-puzzle puzzle)))
    (formatting-table (stream)
      (dotimes (row 4)
	(formatting-row (stream)
	  (dotimes (column 4)
	    (let* ((value (aref puzzle-array row column))
		   (cell-id (encode-puzzle-cell row column)))
	      (updating-output (stream :unique-id cell-id 
				       :cache-value value)
		(formatting-cell (stream :align-x :right)
		  (unless (zerop value)
		    (with-output-as-presentation
			(stream cell-id 'puzzle-cell)
		      (format stream "~2D" value))))))))))))
```

 It creates a grid of table cells from 0 to 15 to layout the `puzzle-array`. There are 16 cells. The display function iterates over these 16 cells, gets the index or `cell-id` (of type `integer`, ranges from 0 to 15), checks its `value` (of type integer, ranges from 0 to 15) and CREATES the presentation of type `puzzle-cell` (of type integer, ignore its range for now) and which displays the `value` on screen. Note that when the `value` is zero, it skips the `with-output-as-presentation`, No presentation object is created.

So for 16 cells, only 15 objects of type `puzzle-cell` is created, because as noted, the `value` zero is skipped. The type of `puzzle-cell` must be `integer` and it range must be from 1 to 15. Each of these `puzzle-cell` objects remember the underlying lisp object `cell-id` (of type integer, ranges from 0 to 15).

Another hint would be the `highlight-presentation` method, which flips the ink for any presentation of type `puzzle-cell` if it gets pointed by mouse. You could see that by pointing mouse to 1 to 15 on display.

And now, about Clouseau.

You should have asked yourself why you couldn't inspect the empty cell (with `cell-id` 15)?

As to why Clouseau shows the `cell-id` when you point to the `puzzle-cell` presentation-type? 

Because in the body of `define-presentation-action`, `object` is bound to the underlying lisp object. If you want to inspect the presentation itself, then modify it as follows (In the image, I clicked on `1`, the object is of type `puzzle-cell`, underlying lisp object is `0`):

![Screenshot from 2020-07-08 13-26-43](https://user-images.githubusercontent.com/10996250/86893257-f7df2f00-c11e-11ea-987c-e562cdfcbc6f.png)

Reply:
I think that my understanding of the concept is just fine. In this application, you manipulate presentations of type puzzle-cell which store objects ranging from 0-15, where you always display only 15 objects, because when the (application logic) cell value is 0, then the cell is not displayed. presentation-type puzzle-cell denotes the object which is the cell id, while its visual representation is drawn with format (and the content of a drawn value has nothing to do with the presentation type puzzle-cell, it has to do with value in the array stored in the application frame).

In other words, what is displayed on the screen, are cell ids, not values from the array puzzle.